examples/erc1155/lib.rs

```diff
@@ -0,0 +1,792 @@
+                                // Our recipient wasn't a smart contract, so there's nothing more for
+                                // us to do
+                                ink_env::debug_println!("Recipient at {:?} from is not a smart contract ({:?})", from, e);
+                            }
+                            _ => {
+                                // We got some sort of error from the call to our recipient smart
+                                // contract, and as such we must revert this call
+                                let msg = ink_prelude::format!(
+                                    "Got error \"{:?}\" while trying to call {:?}",
+                                    e,
+                                    from
+                                );
+                                ink_env::debug_println!("{}", &msg);
+                                panic!("{}", &msg)
+                            }
+                        }
+                    }
+                }
+            }
+
+            Ok(())
+        }
+    }
+
+    impl super::Erc1155 for Contract {
+        #[ink(message)]
+        fn safe_transfer_from(
+            &mut self,
+            from: AccountId,
+            to: AccountId,
+            token_id: TokenId,
+            value: Balance,
+            data: Vec<u8>,
+        ) -> Result<()> {
+            let caller = self.env().caller();
+            if caller != from {
+                ensure!(self.is_approved_for_all(from, caller), Error::NotApproved);
+            }
+
+            ensure!(to != AccountId::default(), Error::ZeroAddressTransfer);
+            self.perform_transfer(from, to, token_id, value, data)?;
+
+            Ok(())
+        }
+
+        #[ink(message)]
+        fn safe_batch_transfer_from(
+            &mut self,
+            from: AccountId,
+            to: AccountId,
+            token_ids: Vec<TokenId>,
+            values: Vec<Balance>,
+            data: Vec<u8>,
+        ) -> Result<()> {
+            let caller = self.env().caller();
+            if caller != from {
+                ensure!(self.is_approved_for_all(from, caller), Error::NotApproved);
+            }
+
+            ensure!(to != AccountId::default(), Error::ZeroAddressTransfer);
+            ensure!(
+                token_ids.len() == values.len(),
+                Error::BatchTransferMismatch,
+            );
+
+            token_ids.iter().zip(values.iter()).for_each(|(&id, &v)| {
+                // If any of the transfers fail we don't want the caller to be able to handle this.
+                // Instead we want to revert the whole call in order to ensure the atomicity of the
+                // batch transfer.
+                assert!(self.perform_transfer(from, to, id, v, data.clone()).is_ok(),
+                    "Failed to transfer {:?} tokens of ID {:?} from {:?} to {:?} as part of a batch transfer.",
+                    v, id, from, to
+                );
```

Comment:
This might be a little controversial, but this is the scenario I want to avoid:

 - Contract wants to transfer `N` tokens 
 - We're only able to transfer `N/2` tokens
 - Contract gets `Err` but ignores it, so `N/2` transfers are written

Reply:
Maybe do a sanity check for all transfers before doing a single actual transfer instead.
That's how I think contracts should generally be written. Checks. Actions. Signals (events).